### ISSUE TITLE
Spark3 Dialect: Support CLUSTER BY in SELECT statement

### ISSUE THREAD

**Issue by R7L208** [sqlfluff/sqlfluff#2484]
## Expected Behaviour
```bash
== [test/fixtures/dialects/spark3/select_cluster_by.sql] FAIL
L:   5 | P:   1 | L003 | Indent expected and not found compared to line #4
==== summary ====
files:             1 violations:        1
clean files:       0 unclean files:     1
avg per file:   1.00 unclean rate:   100%
status:         FAIL
All Finished!
```

## Observed Behaviour
```bash
==== readout ====

=== [ path: test/fixtures/dialects/spark3/select_cluster_by.sql ] ===

INFO       TEMPLATING RAW [jinja] (test/fixtures/dialects/spark3/select_cluster_by.sql)
INFO       Slicing File Template
INFO       [L001] No config_keywords defined for L001
INFO       [L005] No config_keywords defined for L005
INFO       [L006] No config_keywords defined for L006
INFO       [L008] No config_keywords defined for L008
INFO       [L009] No config_keywords defined for L009
INFO       [L015] No config_keywords defined for L015
INFO       [L017] No config_keywords defined for L017
INFO       [L020] No config_keywords defined for L020
INFO       [L021] No config_keywords defined for L021
INFO       [L023] No config_keywords defined for L023
INFO       [L024] No config_keywords defined for L024
INFO       [L025] No config_keywords defined for L025
INFO       [L027] No config_keywords defined for L027
INFO       [L031] No config_keywords defined for L031
INFO       [L032] No config_keywords defined for L032
INFO       [L033] No config_keywords defined for L033
INFO       [L034] No config_keywords defined for L034
INFO       [L035] No config_keywords defined for L035
INFO       [L036] No config_keywords defined for L036
INFO       [L037] No config_keywords defined for L037
INFO       [L039] No config_keywords defined for L039
INFO       [L041] No config_keywords defined for L041
INFO       [L043] No config_keywords defined for L043
INFO       [L044] No config_keywords defined for L044
INFO       [L045] No config_keywords defined for L045
INFO       [L046] No config_keywords defined for L046
INFO       [L048] No config_keywords defined for L048
INFO       [L049] No config_keywords defined for L049
INFO       [L050] No config_keywords defined for L050
INFO       [L051] No config_keywords defined for L051
INFO       [L053] No config_keywords defined for L053
INFO       [L055] No config_keywords defined for L055
INFO       [L056] No config_keywords defined for L056
INFO       [L058] No config_keywords defined for L058
INFO       [L059] No config_keywords defined for L059
INFO       [L060] No config_keywords defined for L060
INFO       [L061] No config_keywords defined for L061
INFO       LEXING RAW (test/fixtures/dialects/spark3/select_cluster_by.sql)
INFO       Elements to Segments.
INFO       Source-only slices: []
INFO       Lexed tokens: ['SELECT', '\n', '    ', 'age', ',', '\n', '    ', 'name', '\n', 'FROM', ' ', 'person', '\n', 'CLUSTER', ' ', 'BY', ' ', 'age', ';', '\n']
INFO       PARSING (test/fixtures/dialects/spark3/select_cluster_by.sql)
INFO       
###
#
# Parsed Tree:
#
###
INFO       
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    base:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'SELECT'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                newline:                                      '\n'
[L:  2, P:  1]      |                whitespace:                                   '    '
[L:  2, P:  5]      |                select_clause_element:
[L:  2, P:  5]      |                    column_reference:
[L:  2, P:  5]      |                        identifier:                           'age'
[L:  2, P:  8]      |                comma:                                        ','
[L:  2, P:  9]      |                newline:                                      '\n'
[L:  3, P:  1]      |                whitespace:                                   '    '
[L:  3, P:  5]      |                select_clause_element:
[L:  3, P:  5]      |                    column_reference:
[L:  3, P:  5]      |                        identifier:                           'name'
[L:  3, P:  9]      |            newline:                                          '\n'
[L:  4, P:  1]      |            [META] dedent:
[L:  4, P:  1]      |            from_clause:
[L:  4, P:  1]      |                keyword:                                      'FROM'
[L:  4, P:  5]      |                whitespace:                                   ' '
[L:  4, P:  6]      |                from_expression:
[L:  4, P:  6]      |                    [META] indent:
[L:  4, P:  6]      |                    from_expression_element:
[L:  4, P:  6]      |                        table_expression:
[L:  4, P:  6]      |                            table_reference:
[L:  4, P:  6]      |                                identifier:                   'person'
[L:  4, P: 12]      |                        newline:                              '\n'
[L:  5, P:  1]      |                        alias_expression:
[L:  5, P:  1]      |                            identifier:                       'CLUSTER'
[L:  5, P:  8]      |                    whitespace:                               ' '
[L:  5, P:  9]      |                    [META] dedent:
[L:  5, P:  9]      |                unparsable:                                   !! Expected: 'Nothing...'
[L:  5, P:  9]      |                    raw:                                      'BY'
[L:  5, P: 11]      |                    whitespace:                               ' '
[L:  5, P: 12]      |                    raw:                                      'age'
[L:  5, P: 15]      |    statement_terminator:                                     ';'
[L:  5, P: 16]      |    newline:                                                  '\n'

INFO       Found unparsable segment...
INFO       [L:  5, P:  9]      |unparsable:                                                   !! Expected: 'Nothing...'
[L:  5, P:  9]      |    raw:                                                      'BY'
[L:  5, P: 11]      |    whitespace:                                               ' '
[L:  5, P: 12]      |    raw:                                                      'age'

INFO       LINTING (test/fixtures/dialects/spark3/select_cluster_by.sql)
== [test/fixtures/dialects/spark3/select_cluster_by.sql] LINTING
INFO       0 -> 'SELECT\n    age,\n    name\nFROM person\nCLUSTER BY age;\n'
CRITICAL   [L031] Applying rule L031 threw an Exception: Fixes must provide an anchor. 
Traceback (most recent call last):
  File "/Users/lorin.dawson/repos/sqlfluff/src/sqlfluff/core/rules/base.py", line 519, in crawl
    res = self._eval(
  File "/Users/lorin.dawson/repos/sqlfluff/src/sqlfluff/rules/L031.py", line 97, in _eval
    self._lint_aliases_in_join(
  File "/Users/lorin.dawson/repos/sqlfluff/src/sqlfluff/rules/L031.py", line 200, in _lint_aliases_in_join
    *[
  File "/Users/lorin.dawson/repos/sqlfluff/src/sqlfluff/rules/L031.py", line 201, in <listcomp>
    LintFix.delete(d)
  File "/Users/lorin.dawson/repos/sqlfluff/src/sqlfluff/core/rules/base.py", line 229, in delete
    return cls("delete", anchor_segment)
  File "/Users/lorin.dawson/repos/sqlfluff/src/sqlfluff/core/rules/base.py", line 144, in __init__
    raise ValueError("Fixes must provide an anchor.")
ValueError: Fixes must provide an anchor.
== [test/fixtures/dialects/spark3/select_cluster_by.sql] FAIL
L:   1 | P:   1 | L031 | Unexpected exception: Fixes must provide an anchor.;
                       | Could you open an issue at
                       | https://github.com/sqlfluff/sqlfluff/issues ? You can
                       | ignore this exception for now, by adding '-- noqa: L031'
                       | at the end of line 1
L:   5 | P:   1 | L003 | Indent expected and not found compared to line #4
L:   5 | P:   1 | L011 | Implicit/explicit aliasing of table.
L:   5 | P:   1 | L014 | Unquoted identifiers must be consistently lower case.
L:   5 | P:   1 | L025 | Alias 'CLUSTER' is never used in SELECT statement.
L:   5 | P:   1 | L029 | Keywords should not be used as identifiers.
L:   5 | P:   9 |  PRS | Line 5, Position 9: Found unparsable section: 'BY age'
==== summary ====
files:             1 violations:        7
clean files:       0 unclean files:     1
avg per file:   7.00 unclean rate:   100%
status:         FAIL
All Finished!
```

## Steps to Reproduce
```bash
> echo "SELECT
    age,
    name
FROM person
CLUSTER BY age;
" > select_cluster_by.sql

> sqlfluff lint --dialect spark3 -vvv select_cluster_by.sql
```

## Dialect
spark3

## Version

```bash
==== sqlfluff ====
sqlfluff:                0.9.2 python:                 3.10.1
implementation:        cpython dialect:                spark3
verbosity:                   3 templater:               jinja
```

## Configuration
```bash
== Raw Config:
core:                                   
    dialect:            spark3              
    disable_noqa:       False               
    encoding:           autodetect          
    ignore:             []                  
    ignore_templated_areas:True                
    nocolor:            False               
    output_line_length: 80                  
    recurse:            True                
    runaway_limit:      10                  
    sql_file_exts:      .sql,.sql.j2,.dml,.ddl
    templater:          jinja               
    verbose:            3                   
indentation:                            
    indented_joins:     False               
    indented_using_on:  True                
    template_blocks_indent:True                
rules:                                  
    allow_scalar:       True                
    comma_style:        trailing            
    indent_unit:        space               
    max_line_length:    80                  
    single_table_references:consistent          
    tab_space_size:     4                   
    unquoted_identifiers_policy:all                 
    Example_L001:                           
        forbidden_columns:  bar, baaz           
    L007:                                   
        operator_new_lines: after               
    L010:                                   
        capitalisation_policy:consistent          
    L011:                                   
        aliasing:           explicit            
    L012:                                   
        aliasing:           explicit            
    L014:                                   
        extended_capitalisation_policy:consistent          
    L016:                                   
        ignore_comment_lines:False               
    L026:                                   
        force_enable:       False               
    L028:                                   
        force_enable:       False               
    L029:                                   
        unquoted_identifiers_policy:aliases             
    L030:                                   
        capitalisation_policy:consistent          
    L038:                                   
        select_clause_trailing_comma:forbid              
    L040:                                   
        capitalisation_policy:consistent          
    L042:                                   
        forbid_subquery_in: join                
    L047:                                   
        prefer_count_0:     False               
        prefer_count_1:     False               
    L052:                                   
        multiline_newline:  False               
        require_final_semicolon:False               
    L054:                                   
        group_by_and_order_by_style:consistent          
    L057:                                   
        additional_allowed_characters:""                  
        allow_space_in_identifier:False               
        quoted_identifiers_policy:all                 
        unquoted_identifiers_policy:all                 
templater:                              
    unwrap_wrapped_queries:True                
    jinja:                                  
        apply_dbt_builtins: True                
        macros:                                 
            dbt_config:         {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
            dbt_is_incremental: {% macro is_incremental() %}True{% endmacro %}
            dbt_ref:            {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
            dbt_source:         {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
            dbt_var:            {% macro var(variable, default='') %}item{% endmacro %}
```

**Comment by R7L208** [comment]
I'm happy to work on a PR for this but would require some initial direction. Namely the `unexpected exception` and if these should be dialect vs rule changes.

**Comment by tunetheweb** [comment]
Looks to me like it’s in the dialect as CLUSTERED rather than CLUSTER. Is that a typo? Or are they different things?

**Comment by R7L208** [comment]
Those are different things. CLUSTERED is used as a parameter when running a CREATE TABLE statement and `CLUSTER` will repartition data in memory as part of a SELECT statement.

https://spark.apache.org/docs/latest/sql-ref-syntax-ddl-create-table-datasource.html#create-datasource-table
https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-clusterby.html

**Comment by tunetheweb** [comment]
Ah sorry. Looked too quickly.

Looks to me like a standard enough dialect improvement. Probably want to implement an override of `SelectStatementSegment` in spark3 dialect and add the new segment in (is it basically the same as `OrderByClauseSegment` as far as syntax is concerned?).

I'd ignore the exception for now. I'm willing to bet once the grammar is sorted, the rule no longer throws that.

Still would be nice not to throw that horrible error, even if the syntax is unparsable as it's a bit scary! Usually means the rule is assuming something will exists and it should be coded better to check that. I'll have a look at that.

**Comment by tunetheweb** [comment]
With 0.9.3 including #2468 this no longer throws the unexpected output. Still needs the syntax support to be added though. Changed the title of the issue.
